### PR TITLE
Experimental middleware to create UI5 integration cards

### PR DESCRIPTION
**Pull request by SRKPendyala**
This PR contains 2 modules,

- `cards-editor-config-writer` is used to update package.json and ui5 yaml configuration file to enable card creation.
- `cards-editor-middleware` has a middleware `sap-cards-generator` which can be use to store cards manifests in the application repository and update i18n files with new texts.

### CODE REVIEW

File: packages/cards-editor-middleware/src/index.ts
Comment (tobiasqueck) [inline packages/cards-editor-middleware/src/index.ts:43]: Please use the `@sap-ux/project-access` module to get the `manifest.json` instead of hardcoding the `webapp` path

File: packages/cards-editor-middleware/src/index.ts
Comment (tobiasqueck) [inline packages/cards-editor-middleware/src/index.ts:25]: please add a short comment explaining what each route handler does

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
* changeset ok
* very good test coverage
* manual testing ok
* add change requests inline

File: packages/cards-editor-config-writer/src/index.ts
Comment (tobiasqueck) [inline packages/cards-editor-config-writer/src/index.ts:27]: Please check this one again because I think you might need `\\"` to actually get the `\"` into your string

Comment (tobiasqueck) [reply packages/cards-editor-config-writer/src/index.ts:27]: I saw that one when looking at the 5 sonar issues reported for this PR. Please address those 5 as well

Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/src/index.ts]: Thanks for pointing this out. We don't need \\ ,  `fiori run --open "test/flpGeneratorSandbox.html#Cards-generator"` is sufficient.

File: packages/cards-editor-middleware/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/index.ts:29]: done

File: packages/cards-editor-config-writer/test/fixtures/lrop-v4/ui5.yaml
Comment (Klaus-Keller) [inline packages/cards-editor-config-writer/test/fixtures/lrop-v4/ui5.yaml:26]: Please remove this internal URL

File: packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/Component.js
Comment (Klaus-Keller) [inline packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/Component.js:1]: Do we need this file for unit testing?

File: packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/localService/metadata.xml
Comment (Klaus-Keller) [inline packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/localService/metadata.xml:1]: Do we need this file for unit testing?

File: packages/cards-editor-middleware/package.json
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/package.json:44]:
```suggestion
```
We removed the mention of `pnpm` in `engines` section of `package.json` on module level and keep it only on root for compatibility reasons, see also PR https://github.com/SAP/open-ux-tools/pull/1688

File: packages/cards-editor-config-writer/package.json
Comment (Klaus-Keller) [inline packages/cards-editor-config-writer/package.json:38]:
```suggestion
```

We removed the mention of `pnpm` in `engines` section of `package.json` on module level and keep it only on root for compatibility reasons, see also PR https://github.com/SAP/open-ux-tools/pull/1688

File: packages/cards-editor-middleware/src/index.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/index.ts:5]: Can we please switch to asynchronous read/write?
Like
```typescript 
import { promises } from 'fs';
..
await promises.writeFile(path, content);
```

File: packages/cards-editor-middleware/src/index.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/index.ts:67]: Please do not hard-code folder `webapp`. Above we correctly use `await getWebappPath()`.

File: packages/cards-editor-middleware/src/index.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/index.ts:102]: Please do not hard-code folder `webapp`. Above we correctly use `await getWebappPath()`.

File: packages/cards-editor-middleware/src/index.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/index.ts:101]: Module `@sap-ux/i18n` can be used to modify i18n files. Just a suggested, we can consider this later.

File: packages/cards-editor-middleware/src/utilities.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/utilities.ts:1]: Please use `promises` for asynchronous file access.

File: packages/cards-editor-middleware/src/utilities.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/utilities.ts:45]: Which `package.json` is read here? I assume the one from this module `@sap-ux/cards-editor-middleware`?
If yes, `../../package.json` points to the wrong location.

 To read the `package.json` we could also just import it, like:
```typescript
import packageJson from '../../package.json';

function getVersion(): string {
    return packageJson.version;
}
```

File: packages/cards-editor-middleware/src/utilities.ts
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/src/utilities.ts:104]: Is this function used anywhere?

File: packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/localService/metadata.xml
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/localService/metadata.xml:1]: Is this file used for unit tests?

File: packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/Component.js
Comment (Klaus-Keller) [inline packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/Component.js:1]: Is this file used for unit tests?

**Review by Klaus-Keller — CHANGES_REQUESTED** [review]
Please see my comments.
Could we also check the 2 new Sonar issues: https://sonarcloud.io/project/issues?id=SAP_open-ux-tools&pullRequest=1668&resolved=false&inNewCodePeriod=true

File: packages/cards-editor-config-writer/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/src/index.ts]: done

File: packages/cards-editor-config-writer/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/src/index.ts]: I have removed it now

File: packages/cards-editor-config-writer/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/src/index.ts]: You are right, its not required. its removed now.

File: packages/cards-editor-middleware/src/utilities.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/utilities.ts]: done.

File: packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/Component.js
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/Component.js:1]: Removed it now

File: packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/localService/metadata.xml
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/test/fixtures/lrop-v4/webapp/localService/metadata.xml:1]: Removed it now

File: packages/cards-editor-middleware/src/utilities.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/utilities.ts]: Removed it now

File: packages/cards-editor-middleware/src/utilities.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/utilities.ts]: Done

File: packages/cards-editor-config-writer/test/fixtures/lrop-v4/ui5.yaml
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/test/fixtures/lrop-v4/ui5.yaml]: Done

File: packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/localService/metadata.xml
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/test/fixtures/lrop-v4/webapp/localService/metadata.xml:1]: Removed it now

File: packages/cards-editor-middleware/package.json
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/package.json]: done

File: packages/cards-editor-middleware/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/index.ts]: done

File: packages/cards-editor-middleware/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-middleware/src/index.ts]: fixed now.

File: packages/cards-editor-config-writer/src/index.ts
Comment (chaudhary1-rohit) [reply packages/cards-editor-config-writer/src/index.ts]: Now I have changed it to major version 0.

**Review by Klaus-Keller — DISMISSED** [review]
Thanks for working on this.
- Code is well structured and easy to understand
- Test coverage is good
- Did a local test

One remark, my proposal would be to add a check for UI5 version >= 1.121, as card generation is not supported with lower releases.

Can also be addressed separate, approved from my side.

**Review by Klaus-Keller — APPROVED** [review]
Re-approving this after version check has been added. Tested version check locally, no issues found.

File: packages/cards-editor-middleware/README.md
Comment (tobiasqueck) [inline packages/cards-editor-middleware/README.md:3]:
```suggestion
A middleware to generate a card manifest and i18n text in the project. Middleware is configured by `@sap-ux/cards-editor-config-writer`
```

File: packages/cards-editor-config-writer/package.json
Comment (tobiasqueck) [inline packages/cards-editor-config-writer/package.json:3]: Please change this to `0.2.0`. In combination with the changeset, it will then be release as `0.3.0` which is important for SAP internal usage so that it has a higher version than the internally released module with version `0.2.6`.

File: packages/cards-editor-middleware/package.json
Comment (tobiasqueck) [inline packages/cards-editor-middleware/package.json:3]: Please change this to `0.2.0`. In combination with the changeset, it will then be release as `0.3.0` which is important for SAP internal usage so that it has a higher version than the internally released module with version `0.2.6`.

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
* changeset good
* suggest fix for readme
* versions need to be changed to have an SAP internal smooth transition
* otherwise, code, tests etc. are all good and it works locally

**Review by tobiasqueck — DISMISSED** [review]

**Review by Klaus-Keller — DISMISSED** [review]
Re-approving after version adjustment

**Review by Klaus-Keller — DISMISSED** [review]

**Review by tobiasqueck — DISMISSED** [review]

**Review by tobiasqueck — APPROVED** [review]

**Comment by tobiasqueck** [comment]
@Klaus-Keller could you please re-approve and also merge right away? Thank you!

**Review by Klaus-Keller — APPROVED** [review]

**Comment by cfg74** [comment]
@tobiasqueck @chaudhary1-rohit I tried to install the card editor for a V2 sample project with `npx @sap-ux/create add cards-editor` and it says:

```
stta.prod.man $npx @sap-ux/create add cards-editor
Executed npm install
```

But no node modules are installed. The `package.json` and `ui5.yaml` file seems to be updated correctly. Can you check/advise?

**Comment by cfg74** [comment]
Additional question, for which project types is this feature available (e.g. FE V2, FE V4, OVP, ALP V2, UI5 freestyle...)?

**Comment by chaudhary1-rohit** [comment]
@cfg74 This feature is available for Object Page V2/V4.

On running `npx @sap-ux/create add cards-editor` cards-editor-middleware should get added to node_modules. I have validated it again and it works. Can you please check again?

**Comment by cfg74** [comment]
@chaudhary1-rohit sorry for the late reply, it still does not work for me. I tried with a standard LROP V2 app. It adds `"@sap-ux/cards-editor-middleware": "0"` to the package.json, and prints out "Executed npm install", but it doesn't do it. This is under Windows with bash, and command `npx --yes @sap-ux/create add cards-editor`. Any idea?

**Comment by cfg74** [comment]
+ @SRKPendyala Can you please check my comment above? Thanks!

**Comment by sivaiitt** [comment]
> @SRKPendyala Can you please check my comment above? Thanks!

We will check it from our side. Could you please check if this command works for you ?
`npm init @sap-ux@latest add cards-editor`

**Comment by cfg74** [comment]
@sivaiitt sorry for the late reply, your suggestion doesn't change the behaviour, npm install is effectively not executed. Did you check on your side?

**Comment by SRKPendyala** [comment]
> > @sivaiitt sorry for the late reply, your suggestion doesn't change the behaviour, npm install is effectively not executed. Did you check on your side?
> 
> @cfg74 : As we discussed in our chat, we've reproduced this issue on a Windows laptop. It appears to be a problem with the "@sap-ux/create" package. We're investigating the issue and will update you once it's resolved.

@cfg74  We raised below pull request to fix create command for Windows. We see that now it fixed `npx @sap-ux/create add cards-editor` for Windows. 

https://github.com/SAP/open-ux-tools/pull/2485

**Comment by SRKPendyala** [comment]
> @cfg74 We raised below pull request to fix `@sap-ux/create add` command for Windows and it is in review. This will fix the issue on Windows.
> 
> #2508

@cfg74  : This issue is fixed now. Could you please help us with the tooling support for project preparation for card generation?

**Pull request merged**
